### PR TITLE
Client steering forced diversity

### PR DESCRIPTION
**Pull request by rawlinp**
## What does this PR (Pull Request) do?

Add a Traffic Router profile parameter that enables the Client Steering Forced Diversity feature. With this feature enabled, Traffic Router will modify its default consistent-hashing behavior w.r.t. CLIENT_STEERING in order to force a more diverse result (one that includes more unique edge caches). This is beneficial for clients to have a better chance at recovery in case of edge cache failures, because they are more likely to retry a different edge cache. With the default behavior, a client can sometimes receive a result with the same edge multiple times, causing a client to potentially retry a failed node multiple times.

- [x] This PR fixes #REPLACE_ME OR is not related to any Issue 

## Which Traffic Control components are affected by this PR?

- Documentation
- Traffic Router

## What is the best way to verify this PR?

1. Run the Traffic Router tests: ```mvn clean verify -Djava.library.path=/usr/local/opt/tomcat-native/lib # replace path with your system's path to tc-native```
2. Manually: create a parameter called `client.steering.forced.diversity` with file = CRConfig.json and value = "false", link it it your Traffic Router profile, and snapshot. Make requests to a CLIENT_STEERING delivery service and observe the default (non-diverse) responses. Now, set the value to "true", snapshot, and make the same requests. Observe that the responses contain a more diverse set of edges for all requests.

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3661/
Test FAILed.

**Comment by mitchell852** [comment]
nice work! any reason why this doesn't just override the default behavior. In other words, why does it have to be enabled via a parameter?

**Comment by rawlinp** [comment]
Thanks! I was mostly for just changing the behavior without requiring a new parameter, but if we don't like the behavior for some reason, this allows us to turn the feature off without having to downgrade. If in the future we think the new behavior should be standard, I guess we can bring it up on the mailing list to remove the parameter and just go forward with the new behavior.

File: traffic_router/core/src/test/java/com/comcast/cdn/traffic_control/traffic_router/core/external/SteeringTest.java [hunk 1]
```diff
@@ -553,4 +556,54 @@ public void itUsesMultiLocationFormatWithMoreThanTwoEntries() throws Exception {
...
+				String[] parts = l.split("\\.");
+				parts[0] = parts[0].replaceFirst("http://", "");
+				actualEdges.add(parts[0]);
+				actualTargets.add(parts[1]);
+			}
+
+			// with client.steering.forced.diversity = true, expect to get all 3 edges back in the locations
+			// by default (client.steering.forced.diversity = false), only one edge would be returned in the locations
+
+			String[] expectedEdgesArray = {"edge-cache-csd-1", "edge-cache-csd-2", "edge-cache-csd-3"};
```
Comment (Daisykh) [inline traffic_router/core/src/test/java/com/comcast/cdn/traffic_control/traffic_router/core/external/SteeringTest.java]: Would it be beneficial to add some more testing on some edge cases? Here's what my initial thoughts were:
- Test on no locations (empty failure case)
- Test on 2 locations available only (I think this one is pretty important)
- Test on different DS weighting values

File: traffic_router/core/src/test/resources/publish/CrStates [hunk 2]
```diff
@@ -146,6 +149,30 @@
       "disabledLocations": [],
       "isAvailable": true
     },
+    "client-steering-diversity-test": {
```
Comment (Daisykh) [inline traffic_router/core/src/test/resources/publish/CrStates]: Not sure if it would be of value to test CS diversity with a case where disabled locations for a target is populated.

Comment (rawlinp) [reply traffic_router/core/src/test/resources/publish/CrStates]: I think the `disabledLocations` list is mostly just an optimization (an unnecessary one IMO but that's a different issue) for Traffic Monitor to tell Traffic Router that entire cachegroups are currently unavailable for a delivery service so that TR can eliminate those cachegroups from its "cache selection process". The CSD stuff isn't really changing that cache selection process since CSD is operating on the _result_ of that process to diversify the end result. So, I'm thinking that test case would add more coverage around the "cache selection process" but not really CSD per se.

File: traffic_router/core/src/test/java/com/comcast/cdn/traffic_control/traffic_router/core/external/SteeringTest.java (hunk 1 above)
Comment (rawlinp) [reply traffic_router/core/src/test/java/com/comcast/cdn/traffic_control/traffic_router/core/external/SteeringTest.java]: So this integration test basically just verifies that CSD provides the expected level of diversity. With 3 available caches and 5 targets, we should expect to see all 3 caches at least once and all 5 targets in the response.

Can you elaborate on what functionality we would expect/verify for those three bullets? They seem to mostly apply to client-steering in general rather than to the expected results of CSD.

File: traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java [hunk 3]
```diff
@@ -537,8 +542,12 @@ public HTTPRouteResult multiRoute(final HTTPRequest request, final Track track)
 			final List<Cache> caches = selectCaches(request, ds, track);
 
 			if (caches != null && !caches.isEmpty()) {
+				if (isClientSteeringDiversityEnabled() && caches.size() > selectedCaches.size()) {
+					caches.removeAll(selectedCaches);
+				}
 				final Cache cache = consistentHasher.selectHashable(caches, ds.getDispersion(), pathToHash);
 				steeringResult.setCache(cache);
+				selectedCaches.add(cache);
```
Comment (ajschmidt) [inline traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java]: Once all of the caches in the coverage zone have been added to the selectedCaches it looks like forced diversity becomes disabled and we revert to the standard consistent hash over all the caches. Is this the desired behavior? I would guess that you would want to clear the selectedCaches and start over.

File: traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java [hunk 4]
```diff
@@ -537,8 +542,12 @@ public HTTPRouteResult multiRoute(final HTTPRequest request, final Track track)
 			final List<Cache> caches = selectCaches(request, ds, track);
 
 			if (caches != null && !caches.isEmpty()) {
+				if (isClientSteeringDiversityEnabled() && caches.size() > selectedCaches.size()) {
+					caches.removeAll(selectedCaches);
```
Comment (ajschmidt) [inline traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java]: Is it safe to assume that the same cache won't appear in the 'caches' list more than once? If so then it seems like you could end up with an empty 'caches' list here.

File: traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java (hunk 3 above)
Comment (rawlinp) [reply traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java]: Yeah, that is the desired behavior that I wanted. We can't really assume that the set of caches assigned to each target is the same (even though in practice that might be true), otherwise I'd think we would do what you said and clear the selectedCaches to start over.

The effect this has is diversifying the result list "enough". E.g. with 3 available caches and 5 targets:
```
edge2.target1
edge1.target2
edge3.target3
<edge 1-3>.target4
<edge 1-3>.target5
```
The last two targets could be any of edges 1 through 3, but by that point we know the result already contains all the possible edges, so by the time we have failed over to target4, either all 3 edges were down, the first 3 target origins were down, or some mix of both. So we've at least covered our bases by trying all 3 possible caches before reaching targets 4 and 5 which could contain duplicates.

File: traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java (hunk 4 above)
Comment (rawlinp) [reply traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java]: Yes, I believe that is safe to assume. `selectCaches` should never return a list with duplicate caches.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3677/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3696/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3706/
Test FAILed.

**Comment by JBevillC** [comment]
I was trying to test this in CiaB and it appears that the both the CZF and deep CZF json file will need to be updated for the integration test to work properly (at least for the correct code path(s) to be tested).  Both the CZF and dCZF json files in CiaB are completely invalid, so all results are based on the maxmind dataset.  Given that CiaB contains almost zero cache group data , it is a lot of work to test this PR in CiaB without writing a script to add a ton of data to TO.

For local testing (outside of CiaB), I haven't tested the latest commits from yesterday, but it does seem necessary to update the deep CZF json file to make sure all code paths are tested with respect to the changes of this PR.  In my testing, it seems just the CZF is being hit and not the dCZF which suggests that the data in the integration test `dczmap.json` file may not be valid.

For example the IPs in `$src/trafficcontrol/traffic_router/core/src/test/resources/dczmap.json` seem to be completely ignored by the traffic router integration test and the fail back is to only return a CZF result (which then gets filtered by the diversity code).

```
{
	"deepCoverageZones":
	{
		"cache-group-01":
		{
			"network6":
			[
				"1234:5678::/64",
				"1234:5679::/64"
			],
			"network":
			[
				"192.168.8.0/24",
				"192.168.9.0/24"
			],
			"caches":
			[
				"host1",
				"host2",
				"host3",
				"host4"
			]
		}
	}
}
```

The only way to test this in CiaB is add a ton of cache groups and associated data with a script and also populate the deep CZF with randomly allocated IP/subnets for each edge cache once they have been created via this same script.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3731/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3815/
Test PASSed.
